middleman_protocol/middleman_protocol/message.py

```diff
@@ -0,0 +1,387 @@
+        pass
+
+    @classmethod
+    def get_frame_format(cls) -> Struct:
+        """ Returns Struct object containing frame structure. """
+        frame_format = Struct(
+            frame_signature=Bytes(FRAME_SIGNATURE_BYTES_LENGTH),
+            signed_part_of_the_frame=Struct(
+                request_id=BytesInteger(FRAME_REQUEST_ID_BYTES_LENGTH),
+                payload_type=Enum(Byte, PayloadType),
+                payload_length=BytesInteger(FRAME_PAYLOAD_LENGTH_BYTES_LENGTH),
+                payload=GreedyBytes,
+            ),
+        )
+
+        assert FRAME_PAYLOAD_LENGTH_STARTING_BYTE == (
+            frame_format.frame_signature.subcon.length +
+            frame_format.signed_part_of_the_frame.request_id.subcon.length +
+            1  # 1 byte for payload_type
+        )
+        assert FRAME_PAYLOAD_STARTING_BYTE == (
+            frame_format.frame_signature.subcon.length +
+            frame_format.signed_part_of_the_frame.request_id.subcon.length +
+            1 +  # 1 byte for payload_type
+            frame_format.signed_part_of_the_frame.payload_length.subcon.length
+        )
+
+        return frame_format
+
+    @classmethod
+    def add_frame_separator(cls, raw_frame: bytes) -> bytes:
+        """ Adds frame separator to raw frame. """
+        assert isinstance(raw_frame, bytes)
+
+        return raw_frame + FRAME_SEPARATOR
+
+    @classmethod
+    def remove_frame_separator(cls, raw_message: bytes) -> bytes:
+        """ Remove frame separator from raw frame. """
+        assert isinstance(raw_message, bytes)
+        assert FRAME_SEPARATOR in raw_message
+
+        frame_separator_index = raw_message.index(FRAME_SEPARATOR)
+
+        assert len(raw_message) == frame_separator_index + FRAME_SEPARATOR_LENGTH
+
+        return raw_message[:frame_separator_index]
+
+    @classmethod
+    def deserialize(
+        cls,
+        raw_message: bytes,
+        public_key: bytes,
+        is_frame_separator_expected: bool=False,
+        is_encoding_expected: bool=False,
+    ) -> Any:
+        """
+        Parses and validates received message in bytes.
+        Returns deserialized frame with deserialized payload.
+        """
+        assert isinstance(is_frame_separator_expected, bool)
+        assert isinstance(is_encoding_expected, bool)
+
+        # Remove frame separator from raw message if it is expected
+        if is_frame_separator_expected:
+            raw_message = cls.remove_frame_separator(raw_message)
+
+        # Reverse encoding on raw message if it is expected
+        if is_encoding_expected:
+            raw_message = cls.escape_decode_raw_message(raw_message)
+
+        # Parse frame
+        frame_format = cls.get_frame_format()
+        try:
+            frame = frame_format.parse(raw_message)
+        except StreamError as exception:
+            raise FrameInvalidMiddlemanProtocolError(f'Deserializing message failed with exception {exception}.')
```

Comment:
Add a colon before `{exception}` or put it in quotes because otherwise you'll get something where it's hard to tell where one message starts and the other ends.

Also I think you should use a more concrete message. What kind of error does `StreamError` represent?

Reply:
From what I have examined, it is thrown when there is not enough bytes to build a message.

Reply:
From what I can see in the [construct core docs](https://construct.readthedocs.io/en/latest/api/core.html):
> StreamError – requested reading negative amount, could not read enough bytes, requested writing different amount than actual data, or could not write all bytes

or
> StreamError – encountered EOF but require is not disabled

or
> StreamError – stream is not seekable and tellable

or
> StreamError – subcon build and encoder transformed more or less than encodeamount bytes, if amount is specified

or
> StreamError – stream failed when reading until EOF

So it looks like it can be considered a client error (invalid input) rather than a bug (library used incorrectly by us). So this would be more accurate:

```
join_messages(f'Protocol frame is malformed and could not be deserialized:', str(exception))
```

It's similar but at least it makes it clear that the problem is caused by invalid data and not by something on our side.